### PR TITLE
Produce events to `content-updated` topic in MSK

### PR DESCRIPTION
**Pull request by bpathak-ons**
### What

* Events are created for each Zebedee content found when the producer flag is switched on
* refactoring to align the package structure with other event driven services

### How to review

* code sense check
* For extended test, switch on the feature flag and check if the event is posted onto content-updated topic

### Who can review

* Anyone

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ type Config struct {
 	HealthCheckCriticalTimeout time.Duration `envconfig:"HEALTHCHECK_CRITICAL_TIMEOUT"`
 	OutputFilePath             string        `envconfig:"OUTPUT_FILE_PATH"`
 	KafkaConfig                KafkaConfig
+	ContentUpdatedTopicFlag    bool `envconfig:"CONTENT_UPDATED_TOPIC_FLAG"`
```
Comment (justinpjose) [inline config/config.go:19]: Can you format this please? You may need to run `make fmt`

Comment (justinpjose) [inline config/config.go:19]: Also, can you reorder this in alphabetical order as well please? It's a good practice to start off with (especially on this new app) and makes it easier to find things

File: config/config.go [hunk 2]
```diff
@@ -58,6 +59,7 @@ func Get() (*Config, error) {
 			ReindexRequestedTopic: "reindex-requested",
 			ContentUpdatedTopic:   "content-updated",
 		},
+		ContentUpdatedTopicFlag: false,
```
Comment (justinpjose) [inline config/config.go:62]: Can you order this alphabetical please?

File: event/producer.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+type ContentUpdatedProducer struct {
+	Marshaller Marshaller
+	Producer   dpkafka.IProducer
+}
+
+// ContentUpdate produce a kafka message for an instance which has been successfully processed.
+func (p ContentUpdatedProducer) ContentUpdate(ctx context.Context, event models.ContentUpdated) error {
+	bytes, err := p.Marshaller.Marshal(event)
+	if err != nil {
+		log.Fatal(ctx, "Marshaller.Marshal", err)
```
Comment (justinpjose) [inline event/producer.go:29]: The description of the log needs to be more clear

```suggestion
		log.Fatal(ctx, "failed to marshal event", err)
```

File: event/producer.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+		return fmt.Errorf(fmt.Sprintf("Marshaller.Marshal returned an error: event=%v: %%w", event), err)
```
Comment (justinpjose) [inline event/producer.go:30]: Firstly, you are already using `Errorf` (https://pkg.go.dev/fmt#Errorf) so you don't need to use `Sprintf` to format the string.

Secondly, you are already logging out the `err` in the previous line so you don't need to return a formatted err. You can just `return err`.

File: event/producer.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+// ContentUpdatedProducer produces kafka messages for instances which have been successfully processed.
+type ContentUpdatedProducer struct {
+	Marshaller Marshaller
+	Producer   dpkafka.IProducer
+}
+
+// ContentUpdate produce a kafka message for an instance which has been successfully processed.
+func (p ContentUpdatedProducer) ContentUpdate(ctx context.Context, event models.ContentUpdated) error {
+	bytes, err := p.Marshaller.Marshal(event)
```
Comment (justinpjose) [inline event/producer.go:27]:
```suggestion
	eventBytes, err := p.Marshaller.Marshal(event)
```
More descriptive this way

File: go.mod [hunk 6]
```diff
@@ -8,6 +8,7 @@ replace github.com/coreos/etcd => github.com/coreos/etcd v3.3.24+incompatible
 require (
 	github.com/ONSdigital/dp-component-test v0.6.5
 	github.com/ONSdigital/dp-healthcheck v1.3.0
+	github.com/ONSdigital/dp-kafka/v2 v2.4.4
```
Comment (cookel2) [inline go.mod:11]: Why has this been changed to an older version of dp-kafka?

**Review by cookel2 — CHANGES_REQUESTED** [review]
Please see question.

File: go.mod (hunk 6 above)
Comment (bpathak-ons) [reply go.mod:11]: Its not changed but picked up with v2.4.4 as last stable release for dp-kafka v2 lib

Comment (cookel2) [reply go.mod:11]: Ok thanks for explaining.

File: config/config.go (hunk 1 above)
Comment (cookel2) [reply config/config.go]: Please can you show me where the ContentUpdatedTopicFlag is being used? I might be missing something but I can't see it being used anywhere so I don't know how to test it.

**Review by cookel2 — CHANGES_REQUESTED** [review]
Please would you mind showing me how to test this. Maybe a demo would be ok?

File: Makefile [hunk 7]
```diff
@@ -9,6 +9,10 @@ LDFLAGS = -ldflags "-X main.BuildTime=$(BUILD_TIME) -X main.GitCommit=$(GIT_COMM
 .PHONY: all
 all: audit test build
 
+.PHONY: fmt
+fmt:
+	go fmt ./...
```
Comment (rahulPalliyalil) [inline Makefile:14]: (suggestion): Not sure if we need this in the makefile. It will be handled by the linter

File: event/producer.go [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+// ContentUpdatedProducer produces kafka messages for instances which have been successfully processed.
+type ContentUpdatedProducer struct {
+	Marshaller Marshaller
+	Producer   dpkafka.IProducer
+}
+
+// ContentUpdate produce a kafka message for an instance which has been successfully processed.
+func (p ContentUpdatedProducer) ContentUpdate(ctx context.Context, event models.ContentUpdated) error {
```
Comment (rahulPalliyalil) [inline event/producer.go:25]: I think it develop this needs to be behind the ```ContentUpdatedTopicFlag``` . Now I can see the flag being created but i don't see a switch for the producer to not publish message when this is turned off based on the flag ? Maybe i might have missed it.

**Review by rahulPalliyalil — CHANGES_REQUESTED** [review]

**Review by rahulPalliyalil — APPROVED** [review]

File: config/config.go (hunk 1 above)
Comment (bpathak-ons) [reply config/config.go]: I have updated the PR including READMe to use this new topic, pls revisit this and let me know if you need further clarifications.

**Review by cookel2 — APPROVED** [review]
looks good to me

**Pull request merged**